### ISSUE TITLE
Add support for the msh file format to ft_read_headshape

### ISSUE THREAD

**Issue by robertoostenveld** [fieldtrip/fieldtrip#2152]
this can be based on the code from https://people.sc.fsu.edu/~jburkardt/m_src/gmsh_io/gmsh_io.html

**Comment by robertoostenveld** [comment]
for this I would need some example data that I will receive through #2150

**Comment by schoffelen** [comment]
The linked code above is not going to be of much help, since simnibs creates 'binary' msh files. This needs a separate reader, for which I started a PR. Also, the binary files exist in different versions, current support is intended for version 2.2, probably we will need version 4 support as well. Note, that @robertoostenveld the test data that @ualsbombe provided, (now on /home/common) is read-only for you (at least the v4 data; I could read the v3 data). Could you change the file permissions please? 

The basic file reader for the v3 file now works, but the interpretation (which I assume comes from the corresponding *.geo file) needs to be incorporated still. Also, we may need to discuss how to represent the different geometries in the headshape (e.g. as separate tet/tri/hex field?, but what then to do with the tissue info)

**Comment by robertoostenveld** [comment]
sorry about that, you should now be able to access all files

```
roboos@mentat001> cd /home/common/matlab/fieldtrip/data/test/original/simnibs
roboos@mentat001> find . -type d -exec chmod 775 {} \;
roboos@mentat001> find . -type f -exec chmod 664 {} \;
```

**Comment by robertoostenveld** [comment]
Regarding representing different geometries: for triangulated surfaces we use a struct-array, like `headshape.bnd` for BEM models, where each `bnd` contains `pos` and `tri`.

For hexahedral and tetrahedral meshes, however, I would not expect that we need multiple meshes. Rather a single mesh, where each volume element maps onto a tissue type, and thereby onto a conductivity value.

Would it not work the same way as https://www.fieldtriptoolbox.org/tutorial/headmodel_eeg_fem/#mesh ?

**Comment by schoffelen** [comment]
Do we intend to call the low-level reader from ft_read_headshape, or ft_read_headmodel?

**Comment by robertoostenveld** [comment]
hmm, good question. When from ft_read_headshape, I would only expect it to contain the geometry. When from ft_read_headmodel, I would expect it to contain tissue types and conductivities. Have you already found that information (i.e. the stuff that complements the geometry).

**Comment by robertoostenveld** [comment]
Yes, it would not surprise me if it is hard-coded

**Comment by schoffelen** [comment]
hmmm, what's your opinion: do we want to be able to deal with generic msh files, or will we go for simnibs support, with some added intelligence?

**Comment by schoffelen** [comment]
or both?

**Comment by robertoostenveld** [comment]
this is also of relevance to @mcvinding

**Comment by robertoostenveld** [comment]
> hmmm, what's your opinion: do we want to be able to deal with generic msh files, or will we go for simnibs support, with some added intelligence?

I propose that ft_read_headSHAPE does it generic, and ft_read_headMODEL includes SIMNIBS specific smartness

**Comment by mcvinding** [comment]
Hello. quick update. In my SimNIBS (v 4.0.0) repository, I have a folder called matlab_tools that contains various functions for reading meshed created with SimNIBS.

This will read mesh:
```Matlab
tst = mesh_load_gmsh4(fullfile(datapath,'ernie.msh'));
```
Output:
```
tst = 

  struct with fields:

              node_data: {}
           element_data: {}
                  nodes: [753953×3 double]
              triangles: [961722×3 int32]
       triangle_regions: [961722×1 int32]
             tetrahedra: [4160581×4 int32]
    tetrahedron_regions: [4160581×1 int32]
```

I guess the next step is to arrange to FT format. Something like this:

```Matlab
tst_mesh = [];
tst_mesh.pos         = tst.nodes;
tst_mesh.tet         = tst.tetrahedra;
tst_mesh.tissue      = tst.tetrahedron_regions;
tst_mesh.tri         = tst.triangles;
```

```Matlab
ft_datatype(tst_mesh)

ans =

    'source+mesh'
```

Though this is only the geometry and not the FEM calculations.

**Comment by schoffelen** [comment]
I found this function, too, thanks @mcvinding. It is however quite non-generic, and specific to the simnibs generated meshes, and a bit ugly. Given that I already started a reader function in the PR, which achieves the same functionality (apart from not being (yet) able to read ASCII formatted .msh files (which simnibs doesn't output afaik), nor version 4 (gmsh version 4, not simnibs v4) .msh files (which simnibs doesn't output), I propose to stick to the version I wrote. In the PR I have added some stub of code to read the mesh for ft_read_headshape. As @robertoostenveld mentioned in an earlier comment: dealing with tissue values should probably go into ft_read_headmodel. 

Should we make an option there to either use the tets/hexs (for FEM) or the tris (possibly for BEM)? Also, note, currently it seems that simnibs produces tetrahedral meshes, as far as I know, the FEMs supported through FT only work with hexahedral meshes.

Do you know of some well documented documentation in simnibs that mentions the interpretation of the tissue types? I recall reading some simnibs generated geo-files from which a human can decode what is what, but I don't know whether this can be made sufficiently generic

**Comment by mcvinding** [comment]
No, I do not know where the tissue type values are documented in the code. But I suppose the values corresponds to Table 2 in this paper: https://doi.org/10.1016/j.neuroimage.2020.117044

**Comment by schoffelen** [comment]
I am not so sure about the ordering (yet), but the tags of the triangles have an offset of 1000, and they represent boundaries between tissue compartments, rather than a tissue of a certain type:

fname4 =

    '/home/common/matlab/fieldtrip/data/test/original/simnibs/v4/ernie.msh'

>> [nodes,elements]=read_gmsh_binary(fname4);
Reading position information for 878741 nodes
Reading element information for 5945957 elements
Reading 1065712 triangles with 2 tags
Reading 4880245 tetrahedra with 2 tags
>> disp(elements)
          triangles: [1065712×3 uint32]
     triangles_indx: [1065712×1 uint32]
      triangles_tag: [1065712×2 uint32]
         tetrahedra: [4880245×4 uint32]
    tetrahedra_indx: [4880245×1 uint32]
     tetrahedra_tag: [4880245×2 uint32]

>> unique(elements.tetrahedra_tag(:,1))'

ans =

  1×6 uint32 row vector

   1   2   3   4   5   6

>> unique(elements.triangles_tag(:,1))'

ans =

  1×7 uint32 row vector

   1001   1002   1003   1004   1005   1006   1007

**Comment by schoffelen** [comment]
well, the boundary meshes in the example data I have look like:
1001 = white matter surface
1002 = pial surface
1003 = inner skull surface
1004 = outer skull surface
1005 = scalp surface
1006 = eye balls (or another pair of almost spherical objects)
1007 = sinuses (air cavities)

**Comment by schoffelen** [comment]
for the tetrahedra:
1 = white matter
2 = grey matter
3 = (I think) csf
4 = bone
5 = skin/soft tissue(?)
6 = pair of spherical objcts

**Comment by mcvinding** [comment]
I had a talk with the SimNIBS devs to explain the values. In the SimNIBS Matlab extension, they are documented in the file `standard_cond.m`  (`/simnibs/matlab_tools/standard_cond.m`). The values are hard coded and meant to be backwards compatible as newer version of SimNIBS added more fine-grained segmentation.

From the file the current values are:
1 = white matter
2 = grey matter
3 = CSF
4 = bone
5 = scalp
6 = eyeballs
7 = compact bone
8 = spongy bone
9 = blood
10 = muscle
100 = electrode rubber (for tDCS)
500 = saline (for tDCS)

This is for the tetrahedra. For triangles in the surfaces, values are 1000 + the value above.

I guess `ft_read_headshape `has to know these values.
